### PR TITLE
Feature/py scenario module

### PR DESCRIPTION
**Pull request by Robert108**
This PR introduces a new module ScenarioModule (SM) for translating openscenario events and parameters to the ATOS world.
Many responsibilities previously handled by the EsminiAdapter now lives in the new module. 

Changes done in this PR include:
* Cmake changes for handling ATOS modules written in Python
* ATOSBase deleted (object ids/names now handled in scenariomodule)
* Remove util.h dependency from EsminiAdapter.
* SM is parameterized with file path for the scenario and which object should be active.
* SM fetches IP config from the vehicle catalog
* SM creates ID mapping from scenario object to ATOS (int) ID. (Esmini also holds an internal object to id mapping, to be solved)
* Delayed start/Triggered start removed from ObjectControl/EsminiAdapter (not needed anyway?)
* Installation scripts can now reinstall all deps (needed when updating esmini version)

### CODE REVIEW

File: atos/CMakeLists.txt [hunk 1]
```diff
@@ -59,6 +63,7 @@ set(WITH_MQTT_BRIDGE ON CACHE BOOL "Enable MQTTBridge module")
 set(WITH_POINTCLOUD_PUBLISHER ON CACHE BOOL "Enable PointcloudPublisher module")
 set(WITH_INTEGRATION_TESTING ON CACHE BOOL "Enable IntegrationTesting module")
 set(WITH_BACK_TO_START ON CACHE BOOL "Enable BackToStart module")
+set(WITH_SCENARIO_MODUEL ON CACHE BOOL "Enable Scenario module")
```
Comment (victorjarlow) [inline atos/CMakeLists.txt]:
```suggestion
set(WITH_SCENARIO_MODULE ON CACHE BOOL "Enable Scenario module")
```

File: atos/common/CRSTransformation.cpp [hunk 2]
```diff
@@ -6,102 +6,130 @@
...
 
-	// Put transformed PJ_POINTS back into TrajPoints
-	for (int i = 0; i < arraySize; i++){
-		trajPoints[i].setXCoord(in[i].xyz.x);
-		trajPoints[i].setYCoord(in[i].xyz.y);
-		trajPoints[i].setZCoord(in[i].xyz.z);
-	}
+  // Put transformed PJ_POINTS back into TrajPoints
+  for (int i = 0; i < arraySize; i++) {
+    if (isnan(in[i].xyz.x)) {
```
Comment (victorjarlow) [inline atos/common/CRSTransformation.cpp]:
```suggestion
    if (isnan(in[i].xyz.x) || isnan(in[i].xyz.y) || isnan(in[i].xyz.z) ) {
```

File: atos/common/CRSTransformation.cpp [hunk 3]
```diff
@@ -6,102 +6,130 @@
...
+      // Apply transformation to the point again. quick fix. Need to understand
```
Comment (victorjarlow) [inline atos/common/CRSTransformation.cpp]:
```suggestion
      // Apply transformation to the point again. quick fix. 
      // TODO: Need to understand why some points are not transformed
```

File: atos/common/CRSTransformation.cpp [hunk 4]
```diff
@@ -6,102 +6,130 @@
...
+      // why some points are not transformed
```
Comment (victorjarlow) [inline atos/common/CRSTransformation.cpp]:
```suggestion
```

File: atos/modules/EsminiAdapter/inc/esminiadapter.hpp [hunk 5]
```diff
@@ -52,9 +53,8 @@ class EsminiAdapter : public Module {
 	static std::unordered_map<uint32_t,std::shared_ptr<ROSChannels::Monitor::Sub>> monrSubscribers;
 	static std::shared_ptr<rclcpp::Service<atos_interfaces::srv::GetObjectTrajectory>> objectTrajectoryService;
 	static std::shared_ptr<rclcpp::Service<atos_interfaces::srv::GetObjectTriggerStart>> startOnTriggerService;
-	static std::shared_ptr<rclcpp::Service<atos_interfaces::srv::GetObjectIp>> objectIpService;
 	static std::shared_ptr<rclcpp::Service<atos_interfaces::srv::GetTestOrigin>> testOriginService;
-	static std::shared_ptr<rclcpp::Service<atos_interfaces::srv::SetObjectIp>> setObjectIpService;
+	rclcpp::Client<atos_interfaces::srv::GetOpenScenarioFilePath>::SharedPtr oscFilePathClient_;	//!< Client to request the current open scenario file path
```
Comment (victorjarlow) [inline atos/modules/EsminiAdapter/inc/esminiadapter.hpp:61]: Why a comment here but not on the services?

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp [hunk 6]
```diff
@@ -36,51 +33,26 @@ using namespace std::chrono_literals;
...
 	v2xPub(*this),
+	storyBoardElementStateChangePub(*this),
 	connectedObjectIdsSub(*this, &EsminiAdapter::onConnectedObjectIdsMessage),
 	exitSub(*this, &EsminiAdapter::onStaticExitMessage),
 	stateChangeSub(*this, &EsminiAdapter::onStaticStateChangeMessage),
 	applyTrajTransform(false),
 	testOriginSet(false)
  {
-	declare_parameter("open_scenario_file","");
+	oscFilePathClient_ = create_client<atos_interfaces::srv::GetOpenScenarioFilePath>(ServiceNames::getOpenScenarioFilePath);
```
Comment (victorjarlow) [inline atos/modules/EsminiAdapter/src/esminiadapter.cpp]: why underscore?

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp [hunk 7]
```diff
@@ -253,93 +198,28 @@ bool EsminiAdapter::isSendDenmAction(const std::string& action)
...
-		return;
-	}
-}
-
 /*!
  * \brief Callback to be executed by esmini when story board state changes.
  * 		If story board element is an action, and the action is supported, the action is run.
  * \param name Name of the StoryBoardElement whose state has changed.
- * \param type Possible values: STORY = 1, ACT = 2, MANEUVER_GROUP = 3, MANEUVER = 4, EVENT = 5, ACTION = 6, UNDEFINED_ELEMENT_TYPE = 0.
+ * \param type Possible values: STORY_BOARD = 1, STORY=2, ACT = 3, MANEUVER_GROUP = 4, MANEUVER = 5, EVENT = 6, ACTION = 7, UNDEFINED_ELEMENT_TYPE = 0.
```
Comment (victorjarlow) [inline atos/modules/EsminiAdapter/src/esminiadapter.cpp]:
```suggestion
 * \param type Possible values: STORY_BOARD = 1, STORY = 2, ACT = 3, MANEUVER_GROUP = 4, MANEUVER = 5, EVENT = 6, ACTION = 7, UNDEFINED_ELEMENT_TYPE = 0.
```

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp [hunk 8]
```diff
@@ -681,38 +559,20 @@ void EsminiAdapter::InitializeEsmini()
...
-	}
-
 
 	// Populate the map tracking Object ID -> esmini index
 	for (int j = 0; j < SE_GetNumberOfObjects(); j++){
 		me->ATOStoEsminiObjectId[std::stoi(SE_GetObjectName(SE_GetId(j)))] = SE_GetId(j);
 	}
 	SE_Close(); // Stop ScenarioEngine
-
+//
```
Comment (victorjarlow) [inline atos/modules/EsminiAdapter/src/esminiadapter.cpp]: ??

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp [hunk 9]
```diff
@@ -726,9 +586,9 @@ void EsminiAdapter::InitializeEsmini()
 		me->pathPublishers.at(id).publish(traj.toPath());		
 
 		// below is for dumping the trajectory points to the console
-		// for (auto& tp : traj.points){
-		// 	RCLCPP_INFO(me->get_logger(), "Trajectory point: %lf, %lf, %lf, %lf, %ld", tp.getXCoord(), tp.getYCoord(), tp.getZCoord(), tp.getHeading(), tp.getTime().count());
-		// }
+		for (auto& tp : traj.points){
+			RCLCPP_INFO(me->get_logger(), "Trajectory point: %lf, %lf, %lf, %lf, %ld", tp.getXCoord(), tp.getYCoord(), tp.getZCoord(), tp.getHeading(), tp.getTime().count());
+		}
```
Comment (victorjarlow) [inline atos/modules/EsminiAdapter/src/esminiadapter.cpp]: Should be commented?

File: conf/conf/atos-param-schema.json [hunk 10]
```diff
@@ -14,15 +14,27 @@
...
+        },
+        "scenario_module":{
+            "ros__parameters": {
+                "active_object_names": {
+                    "type": "string_array",
+                    "items": {
+                        "type": "string"
+                      },
+                    "default": [1],
+                    "description": "String array of object names that should be used in the scenario."
```
Comment (victorjarlow) [inline conf/conf/atos-param-schema.json]:
```suggestion
                    "description": "String array of object names from the vehicle catalog that should be used in the scenario."
```

File: setup_atos.sh [hunk 11]
```diff
@@ -31,24 +31,22 @@ esac
...
     exit 0
+fi
 
 # Only install dependencies
-elif [ "$1" == "-d" ]; then
-    INSTALL_ATOS=false
-else
-    INSTALL_ATOS=true
+if [ "$1" == "-r" ]; then
+    REINSTALL_DEPS="-d"
```
Comment (victorjarlow) [inline setup_atos.sh]:
```suggestion
    REINSTALL_DEPS="-r"
```

File: atos/modules/ScenarioModule/scenariomodule.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+import atos_interfaces.srv
+from os import path
+import rclpy
+import rclpy.logging
+import atos_interfaces.msg
+from rcl_interfaces.msg import SetParametersResult
+from rclpy.node import Node
+from std_msgs.msg import Empty
+from scenariogeneration import xosc
+from storyboard_handler import StoryBoardHandler
```
Comment (victorjarlow) [inline atos/modules/ScenarioModule/scenariomodule.py]:
```suggestion
from modules.ScenarioModule.storyboard_handler import StoryBoardHandler
```

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp (hunk 8 above)
Comment (Robert108) [reply atos/modules/EsminiAdapter/src/esminiadapter.cpp]: Should be removed

File: atos/modules/EsminiAdapter/src/esminiadapter.cpp (hunk 9 above)
Comment (Robert108) [reply atos/modules/EsminiAdapter/src/esminiadapter.cpp]: Yepp! Will fix it!

**Review by victorjarlow — APPROVED** [review]

**Pull request merged**